Add Upgrade guide to docs

kylegach opened the pull request:
## What I did

Add an _Upgrade to Storybook 7_ docs page, heavily based on the [existing guide on Notion](https://chromatic-ui.notion.site/Storybook-7-migration-guide-dbf41fa347304eb2a5e9c69b34503937), and update TOC accordingly.

![](https://user-images.githubusercontent.com/486540/226763143-027a8628-595f-47ba-98b1-4583aaa5173e.png)


## How to test

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/7.0/react/contribute/new-snippets#preview-your-work) for this branch, `kylegach/dx-590-create-migration-docs-route`

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

kylegach committed:
Add Upgrade guide to docs

jonniebigodes commented on docs/upgrade.md:
```diff
@@ -0,0 +1,245 @@
+
+## Major breaking changes
+
+The frontend ecosystem moves quickly, so we’ve accumulated a lot of breaking changes in Storybook 7. The rest of this guide is intended to help you migrate successfully. However, there are a few major breaking changes that you should know about before you go further:
+
```
@kylegach can you polish this section a bit more as the way it's written is basically saying the same thing as what's in the intro above.

jonniebigodes commented on docs/upgrade.md:
````diff
@@ -0,0 +1,245 @@
+2. Add the `legacyMdx1` feature flag to your `.storybook/main.js`:
+
+```js
+export default {
+  features: {
+    legacyMdx1: true,
+  },
+};
````
Can you update this and the other code examples to use the snippet files we currently have? I believe this and the other examples should have a corresponding file. And what's more, it would prevent readers from getting a JS version of the example when they've chosen the TS version in the language selector

kylegach replied:
Thanks for the reminder!

jonniebigodes commented on docs/upgrade.md:
````diff
@@ -0,0 +1,245 @@
+};
+```
+
+Finally, you’ll to add the MDX files to `.storybook/main.js` so that they are indexed:
````
Shouldn't this be "Finally, you'll need to add the MDX files....." or something in that nature?

jonniebigodes reviewed (approved):
This is looking amazing. I left a couple of items that need tweaking and this should be good to go.

jonniebigodes reviewed (changes requested):
I accidentally click the wrong one. This will require some changes before merging.

kylegach committed:
Improve assets

kylegach committed:
Address feedback

- Content updates
- Use snippets when possible

kylegach committed:
Improve snippets consistency, clarity, brevity

kylegach commented:
@jonniebigodes — All comments addressed, I believe. I also added one more commit to improve some of the snippets used on this page.

Finally, I'm almost embarrassed to admit I just now noticed this page: https://storybook.js.org/docs/7.0/react/configure/upgrading. That is now largely redundant with this one. Should we remove it? Do we need to bring any of its content into this one first?

domyen commented:
Would leave the current Upgrade doc and suggest the steps below for the doc in this PR 👇 

**Suggestions for 6.5/X**
- Rename this doc to "Migration guide to 7.X" and reframe the content to specifically target 6»7 migration
- Move this doc to the very top of the sidebar in 6.5 docs

jonniebigodes commented:
@kylegach, don't worry about it. We still have a bit of time to address this. If you're ok with it, why not have a root-level section in the TOC named "Migration guides" (name could be improved) and rename and move this one(upgrade.md) and the Configure / Upgrade page to that section? And at the bottom of that page (configure/upgrade.md), include a section named something like "Learn how to migrate to 7.0" or something of that nature and point to the relevant guide. With it, we'll have a centralized hub where we can continue to expand as new Storybook versions are released (e.g., 7.1, etc.).

kylegach commented:
@jonniebigodes — @domyen and I had a quick chat and settled on the following:

- Leave `/docs/configure/upgrading` as-is
- Change title of new page to _Migration guide for Storybook 7.0_, mostly for SEO reasons and to differentiate from _Upgrading_ title of existing page
- Change slug of new page to `/docs/migration-guide`
- Leave at bottom of TOC on both "latest" and "next" docs
    - We want this guide to be rather prominent, so hiding it in a menu is undesirable
- Update the intro to make it clear which version range is covered: 6.x → 7.0, in this case
    
The end result looks like this:

![](https://user-images.githubusercontent.com/486540/227380076-c8931d40-a17c-41d1-9c8d-6c6d6a6c5278.png)

Dom — We didn't talk about the title used in the TOC. I used _Migrate to 7.0_ to emphasize that the page content is specific to the version being viewed. When `7.0` goes "latest" and `7.1` is "next", the "latest" title would remain the same, and the "next" would update to _Migrate to 7.1_. Cool?

kylegach committed:
Further updates

- Change slug, TOC title, and page title
- Make clear the version range covered

jonniebigodes commented:
@kylegach ok, but there's probably something that we're missing here. For instance, when 7.1 drops, what will happen to this page? And the one related to 7.1? More even, we can not guarantee that users have already migrated to any of those versions. And we'll have to find a suitable place to keep it. Hence why I mentioned that to avoid duplicating work. I'm aware of the emphasis on getting the migration front and center, but in the long run, it will cause us more overhead on maintenance. When we could only have a single point of reference to point our users when they need to migrate/upgrade and keep the maintenance costs at a minimum from our end.

kylegach commented:
That's what I tried to explain in the last paragraph of my most recent comment, but I likely didn't do so very clearly. Let me try again.

Here's what I'm proposing:

- **Today**
    - `6.5` ("latest")
        - `/docs/<framework>/configure/upgrading`
            - TOC: Configure > Upgrading
            - Title: _Upgrading_
    - `7.0` ("next")
        - `/docs/7.0/<framework>/configure/upgrading`
            - TOC, title, and content remain the same as `6.5`
        - `/docs/7.0/<framework>/migration-guide`
            - TOC: Migrate to 7.0
            - Title: _Migration guide for Storybook 7.0_
            - Content: Specific to 6.x → 7.0
- **When `7.0` becomes "latest"**
    - `7.0` ("latest")
        - `/docs/<framework>/configure/upgrading`
            - TOC, title, and content remain the same
        - `/docs/<framework>/migration-guide`
            - TOC: Migrate to 7.0
            - Title: _Migration guide for Storybook 7.0_
            - Content: Specific to 6.x → 7.0
    - `7.1` ("next")
        - `/docs/7.1/<framework>/configure/upgrading`
            - TOC, title, and content remain the same
        - `/docs/7.1/<framework>/migration-guide`
            - TOC: Migrate to 7.1
            - Title: _Migration guide for Storybook 7.1_
            - Content: Specific to 7.0 → 7.1, but starting with a callout pointing to `/docs/7.0/migration-guide` to cover those upgrading from 6.x

This way, each version of the docs contains the most relevant migration guide for that version's users. In general, for a major version, it's previous-major-version → current (e.g. 6.x → 7.0), for a minor version, it's current-major-version → current (e.g. 7.0 → 7.1).

The current and unchanging _Upgrading_ page serves as the general "Here's the upgrade commands" reference and differentiates between upgrading to a stable release vs. a prerelease, which isn't covered by the _Migration guide for Storybook X_ page.
